### ISSUE TITLE
Explanation of check_guest_vmcs_link_pointer_first_word

### ISSUE THREAD

**Issue by connojd** [Bareflank/hypervisor#234]
I'm confused about this part in `vmcs_intel_x64::check_guest_vmcs_link_pointer_first_word`:

``` cpp
...
auto vmcs = g_mm->physint_to_virtptr(vmcs_link_pointer);

if (vmcs == nullptr)
    throw std::logic_error("invalid vmcs physical address");
...
```

Why does it convert `vmcs_link_pointer` to a virtptr and then throw an error if it's null saying it's an invalid physical address? Is it because in that case, `vmcs_link_pointer` is an invalid physical address before the conversion?  Also, why is `physint_to_virtptr` called in the first place? In section 26.3.1.5, the third dash under VMCS link pointer states, 

> The 4 bytes located in memory referenced by the value of the field (as a physical address) must satisfy ...

Why don't we just read from the physical address? Lastly, I think there is an extra 'F' in the statement

``` cpp
auto basic_msr = m_intrinsics->read_msr(IA32_VMX_BASIC_MSR) & 0x7FFFFFFFF;
```

since in this check we only care about the first word.

**Comment by rianquinn** [comment]
There is an extra F so that should be moved. 

When paging is turned on (which it always is with 64bit), you cannot read a physical address. You are always reading from a virtual address (because paging is enabled). 

So basically, we are given a valid physical address, and we need to read it, so we must first get it's virtual address. The reason I say "invalid physical address" is because if the physical address was valid, we should have gotten a virtual address for it. To be more clear we probably could say "invalid physical address, or no physical to virtual mapping", or something like that.

**Comment by connojd** [comment]
What if the guest is unrestricted? Would it then be possible that paging is disabled?

**Comment by rianquinn** [comment]
This code either runs prior to the hypervisor being started (i.e. there are no guests yet), or in the VMM prior to starting the guest itself, so in either case, 64bit is always enabled. 

Now that being said.... I could answer your question a different way. With respect to the guest itself, can paging be disabled if the VMM supports unrestricted guests? The answer is yes. But this code doesn't run in a guest.

**Comment by connojd** [comment]
That makes sense. I assume this code doesn't run even in the event of nested virtualization?

**Comment by rianquinn** [comment]
I like where your heads going.

In the case of nested virtualization, L0 is obviously still 64bit with paging enabled, and L1 is the nested hypervisor, which in the case of Bareflank is also 64bit. We also want to support Hyper-V, and I suspect that it's also 64bit, but might not be. In either case, a hypervisor must have paging enabled (it's once of the requirements), so in this case, even if the VMM is 32bit, paging is still going to be enabled. And then once again, the actual guest which is L2, can be anything. 

But I think all of that misses the point. The code your reviewing is specific to Bareflank, which requires paging (or at least virt -> phys mapping even if they are 1:1), so it doesn't really matter what is going on in the guest. In the case of nested, L0 doesn't case about what is in L1. Now if Bareflank also happens to be L1, then we know this code is still good for the same reasons. 

But yeah, great questions.

**Comment by connojd** [comment]
Fixed in #343
